File: frontend/src/api/antifraud.ts
import apiClient from './client'
import type { ApiResponse } from './types'

export interface RuleDetail {
  rule_name: string
  rule_label: string
  score: number
  max_score: number
  detail: string
  raw_data?: Record<string, unknown>
}

export interface SessionRisk {
  session_id: string
  store_id: string
  table_no: string
  employee_id: string
  employee_name: string
  opened_at: string
  closed_at: string | null
  guest_count: number
  crmeb_order_count: number
  crmeb_total_amount: number
  wework_pay_count: number
  wework_pay_amount: number
  risk_score: number
  risk_level: string
  rules: RuleDetail[]
  is_anomaly: boolean
  anomaly_reason: string | null
  status: string
}

export interface ScanResult {
  date: string
  store_id: string
  total_sessions: number
  anomaly_count: number
  anomaly_rate: number
  high_risk_count: number
  critical_count: number
  sessions: SessionRisk[]
}

export interface AlertListItem {
  session_id: string
  table_no: string
  employee_id: string
  employee_name: string
  risk_score: number
  risk_level: string
  anomaly_reason: string | null
  opened_at: string
  closed_at: string | null
  status: string
  is_anomaly: boolean
}

export interface AlertListData {
  items: AlertListItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface EmployeeStat {
  employee_id: string
  name: string
  anomaly_count: number
  total_sessions: number
  anomaly_rate: number
  avg_risk: number
}

export interface DailyTrend {
  date: string
  total: number
  anomaly_count: number
  anomaly_rate: number
}

export interface AntiFraudStats {
  store_id: string
  date_from: string
  date_to: string
  total_sessions: number
  anomaly_sessions: number
  anomaly_rate: number
  avg_risk_score: number
  by_employee: EmployeeStat[]
  by_rule: { rule_name: string; rule_label: string; avg_score: number; trigger_count: number }[]
  daily_trend: DailyTrend[]
}

export const antifraudAPI = {
  /** 手动触发扫描 */
  scan(date?: string) {
    return apiClient.post<ApiResponse<ScanResult>>(
      '/antifraud/scan',
      { date: date || '' }
    )
  },

  /** 预警列表 */
  getAlerts(params: {
    date_from?: string
    date_to?: string
    risk_level?: string
    employee_id?: string
    page?: number
    page_size?: number
  }) {
    return apiClient.get<ApiResponse<AlertListData>>('/antifraud/alerts', { params })
  },

  /** 预警详情 */
  getAlertDetail(sessionId: string) {
    return apiClient.get<ApiResponse<SessionRisk>>(
      `/antifraud/alerts/${sessionId}`
    )
  },

  /** 防飞单统计 */
  getStats(dateFrom: string, dateTo: string) {
    return apiClient.get<ApiResponse<AntiFraudStats>>('/antifraud/stats', {
      params: { date_from: dateFrom, date_to: dateTo },
    })
  },
}

File: backend/app/schemas/antifraud.py
"""
防飞单模块 Pydantic 模型
- 扫描请求
- 预警记录（基于 table_sessions 的 is_anomaly + anomaly_reason）
- 5 规则明细
- 统计汇总
"""
import uuid
from pydantic import BaseModel, field_validator
from datetime import date as DateType


# ==================== 5 规则明细 ====================

class RuleDetail(BaseModel):
    """单条规则的打分明细"""
    rule_name: str           # cancellation / account_diff / amount_deviation / time_gap / employee_deviation
    rule_label: str          # 作废异常 / 账差 / 金额偏离 / 时间裂隙 / 员工偏差
    score: float             # 0-20
    max_score: float = 20.0
    detail: str              # 触发原因说明
    raw_data: dict | None = None  # 原始数据快照


class SessionRisk(BaseModel):
    """单个开台会话的风险评分结果"""
    session_id: uuid.UUID
    store_id: uuid.UUID
    table_no: str
    employee_id: uuid.UUID
    employee_name: str = ""
    opened_at: str
    closed_at: str | None = None
    guest_count: int
    crmeb_order_count: int
    crmeb_total_amount: float
    wework_pay_count: int
    wework_pay_amount: float
    risk_score: float           # 0-100
    risk_level: str             # low / medium / high / critical
    rules: list[RuleDetail]     # 5 条规则明细
    is_anomaly: bool
    anomaly_reason: str | None = None
    status: str                 # open / closed


class ScanRequest(BaseModel):
    """手动触发扫描"""
    date: str = ""              # YYYY-MM-DD, 空=昨天
    store_id: uuid.UUID | None = None # None = 当前门店

    @field_validator("date")
    @classmethod
    def validate_date(cls, v: str) -> str:
        if not v:
            return v
        import re
        if not re.match(r"^\d{4}-\d{2}-\d{2}$", v):
            raise ValueError("date 格式必须为 YYYY-MM-DD")
        return v


class ScanResponse(BaseModel):
    """扫描结果汇总"""
    date: str
    store_id: uuid.UUID
    total_sessions: int
    anomaly_count: int
    anomaly_rate: float       # 异常率百分比
    high_risk_count: int      # risk_score > 50
    critical_count: int       # risk_score > 80
    sessions: list[SessionRisk]


# ==================== 预警查询 ====================

class AlertFilter(BaseModel):
    """预警列表筛选"""
    date_from: str | None = None
    date_to: str | None = None
    risk_level: str | None = None   # low / medium / high / critical
    employee_id: uuid.UUID | None = None
    status: str | None = None       # open / closed
    page: int = 1
    page_size: int = 20


class AlertListItem(BaseModel):
    """预警列表项（精简）"""
    session_id: uuid.UUID
    table_no: str
    employee_id: uuid.UUID
    employee_name: str
    risk_score: float
    risk_level: str
    anomaly_reason: str | None
    opened_at: str
    closed_at: str | None
    status: str
    is_anomaly: bool


class AlertDetail(SessionRisk):
    """预警详情 = 完整 SessionRisk"""


# ==================== 统计 ====================

class AntiFraudStats(BaseModel):
    """防飞单统计"""
    store_id: uuid.UUID
    date_from: str
    date_to: str
    total_sessions: int
    anomaly_sessions: int
    anomaly_rate: float
    avg_risk_score: float
    by_employee: list[dict]         # [{employee_id, name, anomaly_count, avg_risk}]
    by_rule: list[dict]             # [{rule_name, rule_label, avg_score, trigger_count}]
    daily_trend: list[dict]         # [{date, total, anomaly_count, anomaly_rate}]
